flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1857,7 +1857,7 @@ struct CoordinateOpConversion
     auto c0 = genConstantIndex(loc, lowerTy().indexType(), rewriter, 0);
     mlir::Value base = operands[0];
     auto firTy = coor.getBaseType();
-    mlir::Type cpnTy = getReferenceEleTy(firTy);
+    mlir::Type cpnTy = fir::dyn_cast_ptrOrBoxEleTy(firTy);
```

Comment:
There is one small change here in the failure mode. The `dyn_cast_ptrOrBoxEleTy` will return an unchecked null type here while the old code, at least when assertions are on, would crash.

I think just adding an `assert(cpnTy && "not a reference type");` shores that up.